### PR TITLE
arborx: Enable use of Kokkos from Trilinos for cuda/rocm.

### PR DESCRIPTION
**Pull request by vbrunini**
@spackbot fix style

### CODE REVIEW

**Review by aprokop — CHANGES_REQUESTED** [review]
Thank you for the patch. Indeed, because ArborX package's use of Cuda predates introducing of cuda in the Trilinos package in Spack, we had that limitation.

To finish the patch in the current state, we would also need to remove the following lines
```
    # - current version of Trilinos package does not allow enabling CUDA
    conflicts("+cuda", when="+trilinos")
```

I also see now that starting 14.4, Spack's Trilinos package relies on external Kokkos. We may want to make some changes to address that, but it does not have to be in this PR.

File: var/spack/repos/builtin/packages/arborx/package.py
Comment (vbrunini) [inline var/spack/repos/builtin/packages/arborx/package.py:127]: @aprokop I believe this line is unnecessary now. I know Kokkos required the use of hipcc in the past, but that requirement was removed and for a while now we've been building with amdclang++ directly. I'd like that workflow to be possible (which just deleting these 2 lines enables), but I don't want to break anyone else's workflow. Thoughts on the best approach?

Comment (aprokop) [reply var/spack/repos/builtin/packages/arborx/package.py:127]: That's a good point. I'm not sure. I wonder if we should remove all clauses and just always take the compiler from Kokkos/Trilinos
```
options.append("-DCMAKE_CXX_COMPILER=%s" % spec["kokkos"].kokkos_cxx)
```
or something like that. However, looking at the `kokkos-kernels` spec, it still have the same `hipcc` thingie. It was introduced outside of the development team, however, so I'm not 100% confident it is the right thing to do.
Should ask some people responsible for Kokkos packages.

I'm going to ask some Kokkos developers responsible for Spack maintenance.

File: var/spack/repos/builtin/packages/arborx/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/arborx/package.py]: FWIW. We are switching the core to use f-strings (https://docs.python.org/3/tutorial/inputoutput.html) for string formatting as we make changes so it's something you might want to consider now or in the future.

**Review by aprokop — APPROVED** [review]
I never got an answer from the Kokkos spack team. I think this looks ok, even though I'm not sure how well it plays with the new Trilinos versions with external Kokkos. Let's merge it in, and deal with that later.

File: var/spack/repos/builtin/packages/arborx/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/arborx/package.py]:
```suggestion
            f"-DKokkos_ROOT={kokkos_spec.prefix}",
```

File: var/spack/repos/builtin/packages/arborx/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/arborx/package.py]:
```suggestion
            options.append(f"-DCMAKE_CXX_COMPILER={kokkos_spec.kokkos_cxx}")
```

**Comment by tldahlgren** [comment]
Looks like there's a conflict that needs to be resolved.

**Comment by aprokop** [comment]
I have local conflict resolved, but I can't push into the branch.

**Comment by vbrunini** [comment]
> I have local conflict resolved, but I can't push into the branch.

I've got the "Allow edits and access to secrets by maintainers " box checked for the PR, guess that's not working?

**Comment by aprokop** [comment]
> I've got the "Allow edits and access to secrets by maintainers " box checked for the PR, guess that's not working?

I'm not a maintainer, so no.

I force-pushed to your branch to resolve conflicts.

File: var/spack/repos/builtin/packages/arborx/package.py
Comment (vbrunini) [reply var/spack/repos/builtin/packages/arborx/package.py:127]: Did you get a response from the Kokkos developers on this? Anything else outstanding we need to get this merged?

Comment (aprokop) [reply var/spack/repos/builtin/packages/arborx/package.py:127]: > Did you get a response from the Kokkos developers on this? Anything else outstanding we need to get this merged?

I didn't get the response. It should be fine to merge. I was hoping that would happen after I force pushed, but noone has approved it yet or merged.

I also can't resolve any conversations ¯\_(ツ)_/¯, which also seems to be a blocker.

Comment (vbrunini) [reply var/spack/repos/builtin/packages/arborx/package.py:127]: Looks like I can resolve the conversations so we just need a review, I'll ping slack.

**Review by psakievich — APPROVED** [review]

**Pull request merged**
